checker/src/org/checkerframework/checker/lock/LockVisitor.java

```diff
@@ -4,88 +4,366 @@
+                            }
+                            preconditions.add(Pair.of(lockExpression, LockHeld.class.toString().substring(10 /* "interface " */)));
+                        }
+                    }
+                }
+            }
+        }
+
+        return preconditions;
     }
 
     @Override
     protected void commonAssignmentCheck(AnnotatedTypeMirror varType,
             AnnotatedTypeMirror valueType, Tree valueTree, /*@CompilerMessageKey*/ String errorKey,
-            boolean isLocalVariableAssignement) {
+            boolean isLocalVariableAssignment) {
+
+        Kind valueTreeKind = valueTree.getKind();
+
+        switch(valueTreeKind) {
+            case NEW_CLASS:
+            case NEW_ARRAY:
+                // Avoid issuing warnings for: @GuardedBy(<something>) Object o = new Object();
+                // Do NOT do this if the LHS is @GuardedByBottom.
+                if (!varType.hasAnnotation(GuardedByBottom.class))
+                    return;
+                break;
+            case INT_LITERAL:
+            case LONG_LITERAL:
+            case FLOAT_LITERAL:
+            case DOUBLE_LITERAL:
+            case BOOLEAN_LITERAL:
+            case CHAR_LITERAL:
+            case STRING_LITERAL:
+                // Avoid issuing warnings for: @GuardedBy(<something>) Object o; o = <some literal>;
+                // Do NOT do this if the LHS is @GuardedByBottom.
+                if (!varType.hasAnnotation(GuardedByBottom.class))
+                    return;
+                break;
+            default:
+        }
+
+        boolean skipSubtypeCheck = false;
+
+        // Assigning a value with a @GuardedBy annotation to a variable with a @GuardSatisfied annotation is always
+        // legal. However this is our last chance to check anything before the @GuardedBy information is lost in the
+        // assignment to the variable annotated with @GuardSatisfied. See the Lock Checker manual chapter discussion
+        // on the @GuardSatisfied annotation for more details.
+
+        if (varType.hasAnnotation(GuardSatisfied.class)) {
+            if (valueType.hasAnnotation(GuardedBy.class)) {
+                ExpressionTree tree = (ExpressionTree) valueTree;
 
-        if (valueType.getKind() == TypeKind.NULL) {
-            // Avoid issuing warnings about 'null' not matching the type of the variable.
-            return;
+                checkPreconditions(tree,
+                        TreeUtils.elementFromUse(tree),
+                        tree.getKind() == Tree.Kind.METHOD_INVOCATION,
+                        generatePreconditionsBasedOnGuards(valueType, false),
+                        false);
+
+                skipSubtypeCheck = true;
+            }
+            else if (valueType.hasAnnotation(GuardSatisfied.class)) {
+                // TODO: Find a cleaner, non-abstraction-breaking way to know whether method actual parameters are being assigned to formal parameters.
```

Comment:
As noted on visitMethodInvocation:
If you don't call super in visitMethodInvocation if an error is issued, you won't have to prevent this error.